### DISCUSSION TITLE
Where/How to install dependencies in a Shared Virtual Environment?  Executors are failing to resolve flake8/pytest

### DISCUSSION THREAD

**Discussion by Tungsten78** [lucasvieirasilva/nx-plugins#128] [Q&A] [answered]
I have a single @nxlv library in my repo and ran the migration generator to centralize the dependencies.
* I've installed dependencies at the root via `poetry install`.
* `nx run my-project:lint` and `nx run my-project:test` fail because poetry cannot find the commands flake8 and pytest respectively
* if these are installed globally things seem to work, but I assume the intent is to use the dev dependencies

From the [blog](https://betterprogramming.pub/poetry-python-nx-monorepo-5750d8627024)
> When you use the Shared Virtual Environment approach, you must activate the root virtual environment in your terminal before running any linting, testing, build, add, update, or delete commands.
* not sure I understand how to use `poetry update`, I would like to be able to use `nx affected:lint` and other multi project commands
* at any rate, `poetry update my-project` doesn't seem to change things

I did try with the default isolated dependencies project.
* once I invoked the install executor on the project, the lint and test executors resolve the dependencies ok

disclaimer: never used poetry before today.  Any light that can be shed on what I'm doing wrong would be appreciated.

**Comment by Tungsten78** [comment]
@lucasvieirasilva,

From the blog again,
> When you use the Shared Virtual Environment approach, you must activate the root virtual environment in your terminal before running any linting, testing, build, add, update, or delete commands.

It's not clear to me how to "activate".  Does this refer to installing the dependencies and having poetry establish each .venv?
* Is this best done across the repo with `nx run-many --target install`?

**Comment by lucasvieirasilva** [comment]
Hey @Tungsten78, actually, I updated the `Install Local Dependency` section of the blog, it was referencing the old project generator.

Can you re-run the blog steps again and see if get any other issues?

The `@nxlv/python:poetry-project` is the new version, in the next releases the `@nxlv/python:project` is going to be removed, I just forget to fix one of the references in the blog page.

> It's not clear to me how to "activate". Does this refer to installing the dependencies and having poetry establish each .venv?

it means you have to activate the python virtual environment in your terminal.

There are a lot of ways of doing that, using poetry you can run `poetry shell` in our terminal (in the root of the workspace) and the virtual environment is gonna be activated.

Got it?

**Comment by Tungsten78** [comment]
I thought I had used the new generator but will make double sure.  We used
the single version policy every where so this was the default I attempted.

Regarding activation the concern is with CI.  We clone, run npm install,
then nx affected/runmany.  Currently this works without extra steps with
typescript and dotnet.  Nx-dotnet adds an npm prepare step to call a
restore generator. I think this is an analog to the install executor in
nxlv/python.

Using the non shared mode, I was able to add a “nx run-many -t install” to
have all the python projects setup their venv. Is this equivalent?

**Reply by lucasvieirasilva** [reply]
> Using the non-shared mode, I was able to add a “nx run-many -t install” to
have all the Python projects setup their venv. Is this equivalent?

Correct, although, I don't recommend using non-shared environment, because the installation time might increase based on how many projects you have in the workspace.

I've used the shared venv approach with Azure devops and you only need to activate the venv before the affected command.

You don't need to activate for each project, only in the root before running the command.

Something like:

`source .venv/bin/activate && npx nx affected ...`

Please let me know if that doesn't work in your tests.

**Comment by Tungsten78** [comment]
I can confirm I had been using the correct project generator (no change in behavior)- the issue seems to be the lack of activation.  

My workflow:
1) Open Nx workspace
2) Add nxlv project (my-project) using poetry-project generator
3) Run migrate-to-shared-venv generator
4) `nx lint my-project`  <-- fails to find flake8

If I activate ahead of 4 it works.

What I'm noticing is the activation is scoped (I realize this is the whole point); however, I don't want to change our dev workflow where most devs won't deal directly with the python projects.  Nx abstracts this nicely by hiding the details behind the common targets (lint/test/etc). What I want to avoid is the pre-requisite of having an activation at the root.   

Is there a way to have this persisted?   I would like to embed something in the npm prepare such that post clone / npm i the Nx workspace just works.

**Comment by lucasvieirasilva** [comment]
Hey @Tungsten78, I've created this issue to implement this feature https://github.com/lucasvieirasilva/nx-plugins/issues/132.

**Comment by lucasvieirasilva** [comment]
Hey, @Tungsten78 I've released a new version that supports auto-activate the shared virtual environment as we are discussing in this thread.

Could you check this PR description please: https://github.com/lucasvieirasilva/nx-plugins/pull/138

**Reply by Tungsten78** [reply]
Awesome. Thanks!

**Reply by Tungsten78** [reply]
@lucasvieirasilva I've recently upgraded to v17.0 and migrated to the shared virtual environment (my goal is single versioning)
* dev dependencies were moved to the root `pyproject.toml` and `autoActivate=true` is specified
* my projects also now reference `@nxlv/python:run-commands` for auto-activation
* lastly, I've removed the `nx run-many -t install` I had previously had in `package.json`'s prepare script (as mentioned earlier in this discussion)

The issue is that `nx run my-project:lint` and `nx run my-project:test` are again complaining they cannot find pytest + flake8.

Now with dev dependencies specified at the root, I'm having to run `poetry install` at the root.  Is this expected?

**Reply by lucasvieirasilva** [reply]
@Tungsten78, a couple of things to consider:

1. When you migrate to a shared virtual environment you need to always have your root virtual env up-to-date, if you use the `npx nx run <project>:add` commands to add new packages to the projects the `@nxlv/python` is always going to make sure the root venv is updated, but if you bypass that and install using `poetry install` directly to the project you need to run `poetry lock --no-update` and `poetry install` in the root to keep it updated. 

That's why the projects are referenced in the root pyproject, because poetry needs to know how to resolve the dependencies.

Similar to `tsconfig.base.json` when you use Nx with Node.js

2. It's probably failing because the terminal is using another venv, and the `autoActivate` feature only applies when there's no other venv activated, if you are using vscode terminal, it usually auto-activates the same venv configured in your workspace (it might one of the projects, that doesn't have the dev dependencies anymore)

3. Managing dev dependencies in the root works fine in the way described here, however, if for some reason you need to install a venv in the project (you might need to do that sometime), the dev packages are not going to be present.
   3.1. To solve this problem I recommend you use the `devDependenciesProject` approach.
   
This approach consists of moving all the dev dependencies from the projects to separate projects, this project is referenced in the root `pyproject.toml` and all the local projects as a dev dependency.

Like: 

```
- apps/
  - my-app1
  - my-app2
- libs/
  - shared/
    - development
      - pyproject.toml
```

Where the `shared-development` library contains all the dev dependencies as main dependencies, and you reference in your projects in the dev group.

```shell
npx nx run <project>:add --name shared-development --group dev --local
```
In this way, the `shared-development` is going to be referenced in the root (because all the workspace projects are referenced there), and also in your project in the `dev` group, so, if you upgrade a flake8 version or add another development library all the projects are going to update automatically and the root as well since they are referenced.

The `@nxlv/python` project generator `@nxlv/python:poetry-project` provides a way you point to a devDependencyProject when the project is created, so, you don't need to always run the `add --name shared-development --group dev --local` every time you create a new project.

Reference: https://www.npmjs.com/package/@nxlv/python#devdependenciesproject

Please let me know if have any questions about this approach.

**Reply by Tungsten78** [reply]
Regarding "That's why the projects are referenced in the root pyproject, because poetry needs to know how to resolve the dependencies."

I see the reference from root to project.  
* I was expecting the opposite since run-commands looks to execute poetry in the project directory

To reproduce:
* Create new nx workspace `npx create-nx-workspace@latest nx-python`
* Install plugin `npm install @nxlv/python --save-dev`
* Add project `nx generate @nxlv/python:poetry-project myproject`
* Migrate to shared env `npx nx generate @nxlv/python:migrate-to-shared-venv`
* Execute `nx lint my-project`

Should we expect this to be successful?

**Reply by lucasvieirasilva** [reply]
@Tungsten78 Yes, and I ran the same commands you provided, and it works.

![image](https://github.com/lucasvieirasilva/nx-plugins/assets/12723594/522fda45-990f-4d78-ac9d-a191474f0169)

**Reply by Tungsten78** [reply]
Wow, curious?

I have the same "shared virtual environment detected and no activated, activating...", yet flake8 is missing.
Perhaps another data point, see this ".venv seems to be broken" with `nx test myproject`

![image](https://github.com/lucasvieirasilva/nx-plugins/assets/3805338/43f05fc1-ac27-4040-8cfe-d2a9ebcb86f9)
![image](https://github.com/lucasvieirasilva/nx-plugins/assets/3805338/11a43dba-01ef-409e-a675-e35d4f5dfdd7)

**Reply by Tungsten78** [reply]
Backing up a bit more, the same steps **without** `npx nx generate @nxlv/python:migrate-to-shared-venv`
* `nx lint` produces the same flake8 missing

I am using python 3.11.4 (if it matters).  
I'm having to change 
```  
[tool.poetry.dependencies]
python = ">=3.9,<3.11"
```
to
```
  [tool.poetry.dependencies]
  python = ">=3.9"
```
  
 Sorry for omitting this detail

**Reply by lucasvieirasilva** [reply]
well, the Python version shouldn't matter for this behavior, I've noticed you are using Windows, could you try with Windows WSL? maybe the way I'm activating the venv programmatically doesn't work properly on windows

**Reply by Tungsten78** [reply]
@lucasvieirasilva  - the version seemed to be the issue.   
* There were errors hiding in the command output I hadn't noticed.
* Looks like `npx nx generate @nxlv/python:migrate-to-shared-venv` is what does the initial activation at the root (installing flake8/pytest)

So the steps as indicated now work, however in ci we still need this initial activation (I think):
* while `nx lint my-project` does activate the project, it looks to ignore the root dev dependencies
* whereas, `poetry install` at the root looks to walk the tree activating root + project.   flake8 is found subsequently.

I'll leave `poetry install` in my prepare step.  This works and seems mostly benign.

**Reply by lucasvieirasilva** [reply, marked as answer]
@Tungsten78 got it, well, since you move to a single versioning approach, you always need to run poetry install at the root before running the commands, that's on purpose.
